src/main/scala/viper/silver/ast/utility/QuantifiedPermissions.scala

```diff
@@ -244,6 +244,19 @@ object QuantifiedPermissions {
 
             desugarSourceQuantifiedPermissionSyntax(Forall(vars ++ nestedVars, combinedTriggers, Implies(newCond, nestedRhs)(rhs.pos, rhs.info, rhs.errT))(source.pos,MakeInfoPair(source.info, nested.info),MakeTrafoPair(source.errT,nested.errT)))
 
+          case lt@Let(v, e, bod) => {
+            val forallWithoutLet = Forall(vars, triggers, Implies(cond, bod)(rhs.pos, rhs.info))(source.pos, source.info)
+            // desugar the let-body
+            val desugaredWithoutLet = desugarSourceQuantifiedPermissionSyntax(forallWithoutLet)
+            desugaredWithoutLet.map{
+              case SourceQuantifiedPermissionAssertion(iqp, Implies(icond, irhs)) if (!irhs.isPure) =>
+                // Since the rhs cannot be a let-binding, we expand the let-expression
+                Forall(iqp.variables, iqp.triggers.map(t => t.replace(v.localVar, e)), Implies(icond, irhs.replace(v.localVar, e))(iqp.pos, iqp.info))(iqp.pos, iqp.info)
+              case iforall@Forall(ivars, itriggers, Implies(icond, ibod)) =>
+                // For all pure parts of the quantifier, we just re-wrap the body into a let.
+                Forall(ivars, itriggers, Implies(icond, Let(v, e, ibod)(lt.pos, lt.info))(lt.pos, lt.info))(iforall.pos, iforall.info)
+            }
+          }

```

Comment:
Is it guaranteed that quantifiers that do not match with the first case (`case SourceQuantifiedPermissionAssertion(iqp, Implies(icond, irhs)) if (!irhs.isPure)`) will match with this one? In particular, is it guaranteed that their bodies will be an implication?

Reply:
You cannot see it in the changeset, but the desugaring starts with a check `SourceQuantifiedPermissionAssertion(_, Implies(cond, rhs))` and furthermore, the desugaring always preserves an implication as the body. Therefore, yes, it is guaranteed  that the body is always an implication.